### ISSUE TITLE
SQL inside c# strings syntax highlighting, formatting, code completion and code analysis

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#31028]
<p>Add built in SQL inside c# strings syntax highlighting, code completion and code analysis<strong> </strong>base of connection string for writing raw SQL queries in our code. this feature analysis queries and verify SQL syntax and avoid errors and improvement productivity and when use ExecuteSqlCommand in ef core.</p>

_This issue has been moved from https://developercommunity.visualstudio.com/content/idea/375350/sql-inside-c-strings-syntax-highlighting-formattin.html
VSTS ticketId: 723788_
_These are the original issue comments:_

Mojtaba Kaviani on 11/5/2018, 01:21 AM (2 days ago): <p>Support this feature by jetbrains rider <a target='_blank' href="https://blog.jetbrains.com/dotnet/2018/10/29/sql-inside-c-strings-fragment-editor-run-query-console-language-injection-updates-rider-2018-3/" target="_blank">SQL inside C# strings, fragment editor, run query in console – Language injection updates in Rider 2018.3</a>. please add it to visual studio</p>

Fiona Niu[MSFT] on 11/6/2018, 00:13 AM (37 hours ago): <p>Thank you for taking the time to provide your suggestion. We will do some preliminary checks to make sure we can proceed further.&nbsp;You will hear from us in about a week on our next steps. </p>

**Comment by mojtabakaviani** [comment]
Thank you, i hope add it to Visual Studio very soon...

**Comment by Neme12** [comment]
tagging @CyrusNajmabadi

**Comment by CyrusNajmabadi** [comment]
I'm confused.  This is how i implemented things.  All of this is done through extensions.

**Comment by AgentEnder** [comment]
Any update on this issue? I and several others would love this feature, jetbrains IDE's support it as well.

**Comment by CyrusNajmabadi** [comment]
@AgentEnder i've taken steps toward exposing an extensibility point here for other parties to plug into.  I do not think roslyn woudl ever want to own the sql part here as that would not be a core competency for us, and it would be a huge amount of debt to maintain.

**Comment by AgentEnder** [comment]
It certainly would be a lot to maintain, but I appreciate that steps have been taken towards it. Thanks for the update!

**Comment by CyrusNajmabadi** [comment]
with sql, it wouldn't even make sense to have it baked in since there's no one "sql".  there are all the sql dialects that different sql systems use.  so there would have to be independent impls that were correct for each provider.  For Regex, we can at least point to the canonical .net regex that we're supporting.

**Comment by AgentEnder** [comment]
Yeah, that totally makes sense. There would have to be a way to manage connections and everything.

Once again, appreciate the updates, and that there could be an easy point to extend

**Comment by ToadRage** [comment]
I would love to see this feature implemented

**Comment by CyrusNajmabadi** [comment]
SOmeone will need to supply an actual fully working, full fidelity sql lexer/parser here.

**Comment by ROODAY** [comment]
@CyrusNajmabadi wanted to ping on this again, would a lexer such as this work? https://github.com/bruce-dunwiddie/tsql-parser 

Also found these two:
- https://github.com/Swoogan/tsql-lexer
- https://github.com/JaCraig/SQLParser

**Comment by bricelam** [comment]
https://github.com/microsoft/sqlscriptdom

**Comment by CyrusNajmabadi** [comment]
We coudl use any system that can sit on top of a character abstaction.  Meaning it can parse something that isn't an actual string, but is an abstraction over one.  If people can point to the actual helper functions for lexing, that owuld be great.

**Comment by bricelam** [comment]
I created [a prototype](https://github.com/bricelam/SqlString) using the SQL ScriptDOM library. Here's [where it gets the tokens](https://github.com/bricelam/SqlString/blob/9a09bba53494973fc5793da89812c84f76db7648/Microsoft.AspNetCore.App.Analyzers/SqlEmbeddedLanguage/SqlClassifier.cs#L38-L42).

**Comment by CyrusNajmabadi** [comment]
Dont' use token.ValueText.  Taht won't work for things like `\t` etc in a string.  You want to use the VirtualChar subsystem :)   i can share more info this week.

**Comment by patspoutnik2001** [comment]
yes

**Comment by CyrusNajmabadi** [comment]
Dupe of https://github.com/dotnet/roslyn/issues/30590
